### ISSUE TITLE
Tree: expand/collapse events not symmetrical with 'dynamic="true"'

### ISSUE THREAD

**Issue by kukel** [primefaces/primefaces#3668]
## 1) Environment
- PrimeFaces version: 6.1.x ,6.2.x
- Does it work on the newest released PrimeFaces version? No, Version? Tried 6.2 Elite
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source) 6.2-SNAPSHOT no
- Application server + version: N/A (tried Tomcat 8 and Wildfly 10
- Affected browsers: ALL

## 2) Actual behavior
When a node is opened for the first time, an 'expand' ajax event is fired and the child nodes are populated client side. When the node is closed, a 'collapse' event is send. So far so good. But when the node is opened again, there is no ajax event send. Since the content is present no in the client, it can in a way be understood that there is no ajax event sent. What is strange than is that on closing this node, an collapse event **is** sent. 

When dynamic is false, each open and close results in a new ajax event, even if all content is already present on the client. 

The same behaviour can be achieved with `dynamic="false" when the `cache` attribute, which defaults to true, is set to false. 

## 3) Expected behavior
At least symmetrical behaviour for the expand and collapse events. Two options:
- Only send an ajax event on the initial open and close?
- Like with dynamic="false" keep sending ajax events all the time. 

The latter would be what is most in line with expected behaviour I think. 

## 4) Steps to reproduce

Add `dynamic="true"` to /ui/data/tree/events.xhtml in the showcase

## 5) Sample XHTML

See steps to reproduce (much easier than posting xhtml here

## 6) Sample bean

Use the beans from the /ui/data/tree/events.xhtml case in the showcase. 

## 7) Fix 

If fixing it like stated in 3 would be the way to go, adding  `this.fireExpandEvent(node); ` in line [63 in the tree.js in trunk](https://github.com/primefaces/primefaces/blob/master/src/main/resources/META-INF/resources/primefaces/tree/tree.js#L63)  is what should be done.

## 8) Workaround: 

Add  the following javascript in your page:

```
<h:outputScript>
        
//<![CDATA[

	PrimeFaces.widget.BaseTree.prototype.expandNode = function(node) {
        var $this = this;

        if(this.cfg.dynamic) {
            if(this.cfg.cache && $this.getNodeChildrenContainer(node).children().length > 0) {
                this.showNodeChildren(node);
  	        this.fireExpandEvent(node);
                return;
            }

            if(node.data('processing')) {
                PrimeFaces.debug('Node is already being expanded, ignoring expand event.');
                return;
            }

            node.data('processing', true);

            var options = {
                source: this.id,
                process: this.id,
                update: this.id,
                formId: this.cfg.formId,
                params: [
                    {name: this.id + '_expandNode', value: $this.getRowKey(node)}
                ],
                onsuccess: function(responseXML, status, xhr) {
                    PrimeFaces.ajax.Response.handle(responseXML, status, xhr, {
                            widget: $this,
                            handle: function(content) {
                                var nodeChildrenContainer = this.getNodeChildrenContainer(node);
                                nodeChildrenContainer.append(content);

                                this.showNodeChildren(node);

                                if(this.cfg.draggable) {
                                    this.makeDraggable(nodeChildrenContainer.find('span.ui-treenode-content'));
                                }

                                if(this.cfg.droppable) {
                                    this.makeDropPoints(nodeChildrenContainer.find('li.ui-tree-droppoint'));
                                    this.makeDropNodes(nodeChildrenContainer.find('span.ui-treenode-droppable'));
                                }
                            }
                        });

                    return true;
                },
                oncomplete: function() {
                    node.removeData('processing');
                }
            };

            if(this.hasBehavior('expand')) {
                var expandBehavior = this.cfg.behaviors['expand'];
                expandBehavior.call(this, options);
            }
            else {
                PrimeFaces.ajax.Request.handle(options);
            }
        }
        else {
            this.showNodeChildren(node);
            this.fireExpandEvent(node);
        }
    }
	
	

//]]>
        </h:outputScript>```

**Comment by VsevolodGolovanov** [comment]
If I specify `p:ajax event="expand"` I definitely expect a submit on each expansion.

Verified that the proposed fix works with the current 6.3-SNAPSHOT. PR'ed.

**Comment by VsevolodGolovanov** [comment]
(Fixed indentation problem.)

**Comment by tandraschko** [comment]
Hmm, we really have to discuss this. This would lead to useless ajax requests and you can't differ.
Maybe we need new events.

**Comment by VsevolodGolovanov** [comment]
The documentation clearly says

| Event | Listener Parameter | Fired |
| ------ | ---------- | ------|
| expand | org.primefaces.event.NodeExpandEvent | When a node is expanded. |

Maybe there could be a new event, e.g. `load` for a dynamic tree. But regardless the `expand` event currently doesn't behave like the documentation promises.

**Comment by tandraschko** [comment]
Maybe yes but applying the PR is a performance loss and not required for the most cases. Therefore -1 to just always fire the event.

Whats your usecase?

**Comment by VsevolodGolovanov** [comment]
This doesn't always fire the event. It fires the event only if you have `p:ajax event="expand"`. Same as what already happens for `dynamic="false"` trees.

The use case is preserving the expanded state when switching the tree between different filters or views, essentially.

Come to think of it, you kinda already have a load event - it's when your TreeNode#getChildren gets called.

**Comment by tandraschko** [comment]
@kukel WDYT?

You suggested 2 solutions:

> 1) Only send an ajax event on the initial open and close?
> 2) Like with dynamic="false" keep sending ajax events all the time.

the current PR implements 2) - i personally would prefer 1)

**Comment by tandraschko** [comment]
@VsevolodGolovanov Why do you need the event when cache is enabled?
We could introduce a new flag probably, which is called "fireEventsOnCachedNodes" or something...

**Comment by tandraschko** [comment]
in 6.3, we will have:
expand and collapse events will be fired always, doesn't matter if dynamic true or false
BUT
expand and collapse will both NOT be fired when cache=true; expand will just be fired for the initial request to get the sub-nodes to cache

NOTE: cache=true is the default

**Comment by tandraschko** [comment]
Done, i think everything should work like you expected when you set cache=false

**Comment by VsevolodGolovanov** [comment]
I expect the "expand" and "collapse" events to trigger when I expand and collapse tree nodes...

> Why do you need the event when cache is enabled?

Because I'm interested in "expandness", not "loadness". The requirement is to track, preserve and restore user's expanded state. This is orthogonal to caching.

> We could introduce a new flag probably, which is called "fireEventsOnCachedNodes" or something...

Maybe better make a new "load" event instead. Though again, there is TreeNode#getChildren already...

>applying the PR is a performance loss

It's not. This is what "program to contract, not implementation" is about. If someone coded to a broken behavior, then he introduced a latent bug at that moment. This isn't even a breaking change. Mentioning it in migration guide is plenty.

As is the behavior is confusing. A dynamic cached tree doesn't trigger the collapse event at all. How does this make any sense? Call it "unload" event then.

**Comment by VsevolodGolovanov** [comment]
> applying the PR is a performance loss and not required for the most cases

I've been trying to imagine what those (hypothetical? I doubt there are actual stats) cases are. Is it using the expand event to implement lazy loading between the server and a datasource (database) without making your own TreeNode implementation with a custom getChildren? And the problem is that this lazy loading is implemented to happen unconditionally, relying on expand being triggered once? If so, it can be fixed trivially by adding `if (treeNode.getChildren() == null)`.

I doubt you're talking about the performance of the server-client request itself. Because 1. Why would you listen to the expand event, if you're not actually interested in expanding. If you're interested then you want this request. 2. You're optimizing a case where the user fiddles with the same node for some reason, expanding and collapsing it repeatedly - is this really a common case? I doubt it.
